oai-pmh-handler/src/main/java/no/sikt/nva/oai/pmh/handler/DefaultOaiPmhDataProvider.java

```diff
@@ -116,26 +71,66 @@ private JAXBElement<OAIPMHtype> baseResponse() {
   }
 
   private JAXBElement<OAIPMHtype> listSets() {
-    var response = baseResponse();
-    var value = response.getValue();
+    final ResourceSearchQuery query = buildAllAggregationsQuery();
+
+    var instanceTypes = doSearchAndExtractInstanceTypesFromTypeAggregation(query);
+
+    var uaiResponse = baseResponse();
+    var value = uaiResponse.getValue();
     value.getRequest().setVerb(VerbType.LIST_SETS);
 
     var listSets = objectFactory.createListSetsType();
-    listSets.getSet().addAll(generateSets(objectFactory));
+    listSets.getSet().addAll(generateSets(objectFactory, instanceTypes));
     value.setListSets(listSets);
 
-    return response;
+    return uaiResponse;
+  }
+
+  private Set<String> doSearchAndExtractInstanceTypesFromTypeAggregation(
+      ResourceSearchQuery query) {
+    final HttpResponseFormatter<ResourceParameter> response;
+    try {
+      response = query.doSearch(opensearchClient);
+    } catch (RuntimeException e) {
```

Comment:
Is this in fact an ApiGatewayException?

Reply:
No, it is a result of the error handling done here: https://github.com/BIBSYSDEV/nva-search-api/blob/main/search-commons/src/main/java/no/unit/nva/search/common/OpenSearchClient.java#L77